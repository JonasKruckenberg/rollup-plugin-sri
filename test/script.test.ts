import { getHtml, emitAsset } from './util'
import { rollup } from 'rollup'
import { join } from 'path'
import sri from '../index'

process.chdir(join(__dirname, 'fixtures'));

describe('script', () => {
  test('basic', async () => {
    const bundle = await rollup({
      input: '1.js',
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
              <script src="1.js"></script>
          </body>
        </html>`),
        sri()
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
  })

  test('inactive', async () => {
    const bundle = await rollup({
      input: '1.js',
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
              <script src="1.js"></script>
          </body>
        </html>`),
        sri({ active: false })
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
  })

  test('no src', async () => {
    const bundle = await rollup({
      input: '1.js',
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
              <script></script>
          </body>
        </html>`),
        sri()
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
  })

  test('remote', async () => {
    const bundle = await rollup({
      input: '1.js',
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
              <script src="https://raw.githubusercontent.com/JonasKruckenberg/rollup-plugin-sri/master/test/fixtures/1.js"></script>
          </body>
        </html>`),
        sri()
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
  })

  test('multiple hash algorithms', async () => {
    const bundle = await rollup({
      input: '1.js',
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
              <script src="1.js"></script>
          </body>
        </html>`),
        sri({ algorithms: ['sha256', 'sha384', 'sha512', 'whirlpool'] })
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
  })

    test('uppercase hash algorithms', async () => {
      const bundle = await rollup({
        input: '1.js',
        plugins: [
          emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
              <script src="1.js"></script>
          </body>
        </html>`),
          sri({ algorithms: ['SHA256', 'SHA384', 'SHA512'] })
        ]
      })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
  })

  test('invalid hash algorithm', async () => {
    const onwarn = jest.fn()

    const bundle = await rollup({
      input: '1.js',
      onwarn,
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
              <script src="1.js"></script>
          </body>
        </html>`),
        sri({ algorithms: ['sha1'] })
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
    expect(onwarn.mock.calls).toHaveLength(1)
    expect(onwarn.mock.calls[0][0]).toHaveProperty('message','Insecure hashing algorithm "sha1" will be rejected by browsers!')
  })

  test('crossorigin attribute', async () => {
    const bundle = await rollup({
      input: '1.js',
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
              <script src="1.js"></script>
          </body>
        </html>`),
        sri({ crossorigin: 'use-credentials' })
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
  })

  test('invalid link', async () => {
    const onwarn = jest.fn()

    const bundle = await rollup({
      input: '1.js',
      onwarn,
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
              <script src="invalid.js"></script>
          </body>
        </html>`),
        sri({ crossorigin: 'use-credentials' })
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
    expect(onwarn.mock.calls).toHaveLength(1)
    expect(onwarn.mock.calls[0][0]).toHaveProperty('message','could not resolve resource "invalid.js"!')  
  })

  test('nointegrity attribute', async () => {
    const bundle = await rollup({
      input: '1.js',
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
              <script data-nointegrity="" src="1.js"></script>
          </body>
        </html>`),
        sri()
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
  })
})