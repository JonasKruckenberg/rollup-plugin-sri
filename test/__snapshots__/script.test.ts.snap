// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`script basic 1`] = `
Object {
  "fileName": "index.html",
  "source": "<!DOCTYPE html><html lang=\\"en\\"><head></head><body>
              <script src=\\"1.js\\" integrity=\\"sha384-hxLOTPNeroviduGwj6Po7ksKHF0YRfwVZGjcsAnyJGlTHnQFW/268mrNS7EhSMvT\\" crossorigin=\\"anonymous\\"></script>
          
        </body></html>",
}
`;

exports[`script crossorigin attribute 1`] = `
Object {
  "fileName": "index.html",
  "source": "<!DOCTYPE html><html lang=\\"en\\"><head></head><body>
              <script src=\\"1.js\\" integrity=\\"sha384-hxLOTPNeroviduGwj6Po7ksKHF0YRfwVZGjcsAnyJGlTHnQFW/268mrNS7EhSMvT\\" crossorigin=\\"use-credentials\\"></script>
          
        </body></html>",
}
`;

exports[`script invalid hash algorithm 1`] = `
Object {
  "fileName": "index.html",
  "source": "<!DOCTYPE html><html lang=\\"en\\"><head></head><body>
              <script src=\\"1.js\\" integrity=\\"sha1-WLJ8+6KFXHup1DH0hA2yfQhSqvM=\\" crossorigin=\\"anonymous\\"></script>
          
        </body></html>",
}
`;

exports[`script invalid link 1`] = `
Object {
  "fileName": "index.html",
  "source": "<!DOCTYPE html><html lang=\\"en\\"><head></head><body>
              <script src=\\"invalid.js\\"></script>
          
        </body></html>",
}
`;

exports[`script multiple hash algorithms 1`] = `
Object {
  "fileName": "index.html",
  "source": "<!DOCTYPE html><html lang=\\"en\\"><head></head><body>
              <script src=\\"1.js\\" integrity=\\"sha256-4Wu1uCLo1qUmpY7BjzSoj9gZnYvMLQsg0tFS3ZEUiT4= sha384-hxLOTPNeroviduGwj6Po7ksKHF0YRfwVZGjcsAnyJGlTHnQFW/268mrNS7EhSMvT sha512-QZarufP+9NF0jA1Grhr2cRmLGcceCda8LW2RRHfn3N8P/pxEo+yovxc/RhBJ1Vfuo85A7KP/AZe9monwbbR1TA== whirlpool-s8ZeDZB31uI9Rpj8gH/ivY03VM1eLI3ZgBJzC/030190Z9bi8ixzCYTdKHQPM1Yu7xcKSlIIOYlu594qc0C20g==\\" crossorigin=\\"anonymous\\"></script>
          
        </body></html>",
}
`;

exports[`script remote 1`] = `
Object {
  "fileName": "index.html",
  "source": "<!DOCTYPE html><html lang=\\"en\\"><head></head><body>
              <script src=\\"https://raw.githubusercontent.com/JonasKruckenberg/rollup-plugin-sri/master/test/fixtures/1.js\\"></script>
          
        </body></html>",
}
`;

exports[`script uppercase hash algorithms 1`] = `
Object {
  "fileName": "index.html",
  "source": "<!DOCTYPE html><html lang=\\"en\\"><head></head><body>
              <script src=\\"1.js\\" integrity=\\"sha256-4Wu1uCLo1qUmpY7BjzSoj9gZnYvMLQsg0tFS3ZEUiT4= sha384-hxLOTPNeroviduGwj6Po7ksKHF0YRfwVZGjcsAnyJGlTHnQFW/268mrNS7EhSMvT sha512-QZarufP+9NF0jA1Grhr2cRmLGcceCda8LW2RRHfn3N8P/pxEo+yovxc/RhBJ1Vfuo85A7KP/AZe9monwbbR1TA==\\" crossorigin=\\"anonymous\\"></script>
          
        </body></html>",
}
`;
